validator_client/src/graffiti_file.rs

```diff
@@ -0,0 +1,164 @@
+use serde_derive::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::{prelude::*, BufReader};
+use std::path::PathBuf;
+use std::str::FromStr;
+
+use bls::blst_implementations::PublicKey;
+use types::{graffiti::GraffitiString, Graffiti};
+
+#[derive(Debug)]
+pub enum Error {
+    InvalidFile(std::io::Error),
+    InvalidLine,
+    InvalidPublicKey(String),
+    InvalidGraffiti(String),
+}
+
+/// Struct to load validator graffitis from file.
+/// The graffiti file is expected to have the following structure
+///
+/// default: Lighthouse
+/// public_key1: graffiti1
+/// public_key2: graffiti2
+/// ...
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct GraffitiFile {
+    graffiti_path: PathBuf,
+    graffitis: HashMap<PublicKey, Graffiti>,
+    default: Option<Graffiti>,
+}
+
+impl GraffitiFile {
+    pub fn new(graffiti_path: PathBuf) -> Self {
+        Self {
+            graffiti_path,
+            graffitis: HashMap::new(),
+            default: None,
+        }
+    }
+
+    /// Loads the graffiti file and populates the default graffiti and `graffitis` hashmap.
+    /// Returns the graffiti corresponding to the given public key if present, else returns the
+    /// default graffiti.
+    ///
+    /// Returns an error if loading from the graffiti file fails.
+    pub fn load_graffiti(&mut self, public_key: &PublicKey) -> Result<Option<Graffiti>, Error> {
+        self.read_graffiti_file()?;
+        Ok(self.graffitis.get(public_key).copied().or(self.default))
+    }
+
+    /// Reads from a graffiti file with the specified format and populates the default value
+    /// and the hashmap.
+    ///
+    /// Returns an error if the file does not exist, or if the format is invalid.
+    pub fn read_graffiti_file(&mut self) -> Result<(), Error> {
+        let file = File::open(self.graffiti_path.as_path()).map_err(Error::InvalidFile)?;
+        let reader = BufReader::new(file);
+
+        let lines = reader.lines();
+
+        for line in lines {
+            let line = line.map_err(|_| Error::InvalidLine)?;
+            let (pk_opt, graffiti) = read_line(&line)?;
+            match pk_opt {
+                Some(pk) => {
+                    self.graffitis.insert(pk, graffiti);
+                }
+                None => self.default = Some(graffiti),
+            }
+        }
+        Ok(())
+    }
+}
+
+/// Parses a line from the graffiti file.
+///
+/// `Ok((None, graffiti))` represents the graffiti for the default key.
+/// `Ok((Some(pk), graffiti))` represents graffiti for the public key `pk`.
+/// Returns an error if the line is in the wrong format or does not contain a valid public key or graffiti.
+fn read_line(line: &str) -> Result<(Option<PublicKey>, Graffiti), Error> {
+    if let Some(i) = line.find(':') {
+        let (key, value) = line.split_at(i);
+        if value.is_empty() {
```

Comment:
I noticed some interesting behaviour with these two lines which is suspect is a side-effect of protecting against the array access on L87:

- `default:` (no trailing space): Triggers an `Error::InvalidLine`.
- `default: ` (with trailing space): Is accepted and results in empty graffiti.

I feel that we probably desire these two things:

1. A strictly-enforced schema which requires a colon and space (`: `).
1. The ability to allow users to provide empty graffiti.

Reply:
> `default: (no trailing space): Triggers an Error::InvalidLine`

This shouldn't return `InvalidLine` as `value` in this case should be the string `:`. So when we do `value[1..].trim()`, we get an empty string which would get parsed as an empty `GraffitiString`.

> A strictly-enforced schema which requires a colon and space (: )

I didn't think the space was necessary, so had it like this. Can enforce this if required :)

Reply:
Oh you're totally right, sorry my bad.